### PR TITLE
#63 Don’t cast from an interface to a concrete type

### LINKED ISSUE #63: Rule: Don’t cast from an interface to a concrete type
Replaces http://jira.sonarsource.com/browse/LINTVS-170

Whenever you need to cast from an interface to a concrete type that means that something is wrong with the abstractions you are using, something is missing from the interface. So you should fix the interface instead of casting, because casting can throw an exception during runtime. 

Attachment: https://gist.github.com/dbolkensteyn/0ac2cc784069eb1526e7

Implements http://jira.sonarsource.com/browse/RSPEC-3215

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #63

### CODE REVIEW

**Comment by dbolkensteyn** [comment]
@tvsonar Can you please rebase this one before I test it? (to avoid running into the issues due to the project renaming)

**Comment by tvsonar** [comment]
@dbolkensteyn Done.

File: src/SonarLint/Rules/CastConcreteTypeToInterface.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                {
+                    var castExpression = (BinaryExpressionSyntax)c.Node;
+                    var castedTo = c.SemanticModel.GetTypeInfo(castExpression.Right).Type;
+                    var castedFrom = c.SemanticModel.GetTypeInfo(castExpression.Left).Type;
+                    ReportIssue(castedTo, castedFrom, c);
+                },
+                SyntaxKind.AsExpression);
+        }
+
+        public static void ReportIssue(ITypeSymbol castedTo, ITypeSymbol castedFrom,
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/CastConcreteTypeToInterface.cs]: @tvsonar the method name here is not ideal: the method doesn't just report an issue as its name indicates, but it actually does it conditionally - it therefore would make more sense to call it `CheckSomething()`

File: src/SonarLint/Rules/CastConcreteTypeToInterface.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    [Tags("design")]
+    public class CastConcreteTypeToInterface : DiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3215";
+        internal const string Title = "\"interface\" instances should not be cast to concrete types";
+        internal const string Description =
+            "Needing to cast from an interface to a concrete type indicates that something is wrong with the abstractions in use, " +
+            "likely that something is missing from the interface. Instead of casting to a discrete type, the missing functionality " +
+            "should be added to the interface. Otherwise there is the risk of runtime exceptions.";
+        internal const string MessageFormat = "Remove this cast and edit the \"interface\" to add the missing functionality.";
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/CastConcreteTypeToInterface.cs]: @tvsonar this issue message is strange: "interface" should be written without the quotes around it - else it sounds like we're talking about a fake interface...

File: src/SonarLint/Rules/CastConcreteTypeToInterface.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        internal const string Category = "SonarQube";
+        internal const Severity RuleSeverity = Severity.Major;
+        internal const bool IsActivatedByDefault = false;
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/CastConcreteTypeToInterface.cs]: @tvsonar I've updated the rule to be enabled by default as discussed

**Comment by dbolkensteyn** [comment]
LGTM - but see the comments

**Pull request merged**